### ISSUE TITLE
Nested Endpoints?

### ISSUE THREAD

**Issue by SingleShot** [dipscope/JsonApiEntityProvider.TS#2]
Hi. I've started using this very well-crafted project but am unable to figure out how to support nested resources.

For example, if an endpoint is in the pattern `/books/:id/chapters` and the type is `chapter`, how would the store or resource be configured for this?

Thanks.

**Comment by dpimonov** [comment]
Hi. Thanks for the question.

You can reference a spec file for currently available use cases [here](https://github.com/dipscope/JsonApiEntityProvider.TS/blob/main/spec/json-api-entity-provider.spec.ts). Basically from library API perspective you should only specify a resource type and then use a fluent API to browse, include, update and delete your resources.

For your concrete case - if you have 2 resources:

```
@Type()
@JsonApiResource({
    type: 'chapters'
})
class Chapter
{
    @Property(Number) public id: number;
    @Property(() => Book) public book: Book;
}

@Type()
@JsonApiResource({
    type: 'books'
})
class Book
{
    @Property(Number) public id: number;
    @Property(EntityCollection, [() => Chapter]) public chapters: EntityCollection<Chapter>;
}
```

Then related entities can be loaded using [include api](https://jsonapi.org/format/1.0/#fetching-includes) like this:

```
const bookId = 123;
const bookWithChapters = await bookSet.includeCollection(b => b.chapters).find(bookId);
```

If you do not have include api enabled then another option might be a [filtering api](https://jsonapi.org/format/1.0/#fetching-filtering).

```
const bookId = 123;
const bookChapters = await chapterSet.filter((c, fe) => fe.eq(c.book.id, bookId)).findAll();
```

As filtering api might differ between servers there is an [interface available](https://github.com/dipscope/JsonApiEntityProvider.TS/blob/main/src/json-api-filter-expression-visitor.ts) in the options which you can implement and specify how filter expression should be transformed to an actual request to the server. There is currently only [implementation](https://github.com/dipscope/JsonApiEntityProvider.TS/blob/main/src/filter-expression-visitors/json-api-net-filter-expression-visitor.ts) for [JsonApiNet](https://www.jsonapi.net/) but you can easily implement your own with some custom logic. There is a whole metadata tree for your resources available provided by entity store for each resource and property. Unfortunatelly there is no documentation for that but implementation will not differ much from existing one.

There are also plans to implement lazy loading API. So for example if you have a book resource in memory so request can be done to `/books/:id/chapters` endpoint simply by accessing a `book.chapters` property. But so far I have not got a time slot to think and implement this feature. So as for now there are no other ways except 2 provided above to access related resource data.

I hope I answered your question. If you have any further questions feel free to ask them.

**Comment by dpimonov** [comment]
Included in v2.1.0.
